### ISSUE TITLE
CA1508 false positive in simple nullable code

### ISSUE THREAD

**Issue by yaakov-h** [dotnet/roslyn-analyzers#5747]
### Analyzer

**Diagnostic ID**: CA1508: Avoid Dead Conditional Code

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 6.0.0 (Latest)

### Describe the bug

See code example below, CA1508 fires when a nullable local is checked to see if it is null or not.

The VS IDE correctly infers that the type of the local variable is a nullable string, `string?`:

![image](https://user-images.githubusercontent.com/426009/142977976-cff47659-da9a-4422-8875-6bb6fd6dacdd.png)

### Steps To Reproduce

Compile the following code:

```csharp
namespace ClassLibrary1
{
    public class Class1
    {
        public bool TryFoo()
        {
            if (Accept(MyVisitor.Instance) is var fieldIdentifier && fieldIdentifier != null)
            {
                return true;
            }

            return false;
        }

        static TResult Accept<TResult>(Visitor<TResult> visitor) => default;

        class Visitor<TResult>
        {
        }

        class MyVisitor : Visitor<string?>
        {
            public static MyVisitor Instance { get; } = new MyVisitor();
        }
    }
}
```

### Expected behavior

No CA1508 warnings.

### Actual behavior

> Class1.cs(7,70,7,93): error CA1508: 'fieldIdentifier != null' is always 'true'. Remove or refactor the condition(s) to avoid dead code.

**Comment by Youssef1313** [comment]
@mavasani The number of FP reports on CA1508 is increasing. IMO this really worth a look.

**Comment by Evangelink** [comment]
It seems that the new pattern is implemented in the dataflow but the engine is learning the wrong information. In the case mentioned above `fieldIdentifier` can be `null` and so the check is needed but replacing the line with ` if (Accept(MyVisitor.Instance) is {} fieldIdentifier && fieldIdentifier != null)` results in no diagnostic while in this case the check for non-null is redundant.
